codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+                        }
+                    }
+                }
+            }
+        }
+        rust("#T(tokens)?", nestedParser)
+    }
+
+    private fun RustWriter.deserializeUnion(shape: UnionShape) {
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val symbol = symbolProvider.toSymbol(shape)
+        val nestedParser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                *codegenScope,
+                "Shape" to symbol
+            ) {
+                rust("let mut variant = None;")
+                rustBlock("match tokens.next().transpose()?") {
+                    rustBlockTemplate(
+                        """
+                            Some(#{Token}::ValueNull { .. }) => return Ok(None),
+                            Some(#{Token}::StartObject { .. }) =>
+                        """,
+                        *codegenScope
+                    ) {
+                        objectKeyLoop {
+                            rustTemplate(
+                                """
+                                if variant.is_some() {
+                                    return Err(#{Error}::custom("encountered mixed variants in union"));
+                                }
+                                """,
+                                *codegenScope
+                            )
+                            withBlock("variant = match key.as_escaped_str() {", "};") {
+                                for (member in shape.members()) {
+                                    val variantName = member.memberName.toPascalCase()
+                                    rustBlock("${member.wireName().dq()} =>") {
+                                        withBlock("Some(#T::$variantName(", "))", symbol) {
+                                            deserializeMember(member, false)
+                                        }
+                                    }
+                                }
+                                // TODO: Handle unrecognized union variants
+                                rust("_ => None")
+                            }
+                        }
+                    }
+                    rustTemplate(
+                        "_ => return Err(#{Error}::custom(\"expected start object or null\"))",
+                        *codegenScope
+                    )
+                }
+                rust("return Ok(variant);")
+            }
+        }
+        rust("#T(tokens)?", nestedParser)
+    }
+
+    private fun RustWriter.objectKeyLoop(inner: RustWriter.() -> Unit) {
+        rustBlock("loop") {
+            rustBlock("match tokens.next().transpose()?") {
```

Comment:
it may be helpful to offer a wrapper around `Peekable` inside of smithy-json that has a nicer interface to avoid all the transposition etc.